// @ibm-cloud/event-notifications-node-admin-sdk/event-notifications/v1.js
"use strict";
/**
 * (C) Copyright IBM Corp. 2025.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
/**
 * IBM OpenAPI SDK Code Generator Version: 3.70.0-7df966bf-20230419-195904
 */
/* eslint-disable max-classes-per-file */
/* eslint-disable no-await-in-loop */
var extend = require("extend");
var ibm_cloud_sdk_core_1 = require("ibm-cloud-sdk-core");
var common_1 = require("../lib/common");
var ibm_cloud_sdk_core_2 = require("ibm-cloud-sdk-core");
/**
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * API Version: 1.0
 */
var EventNotificationsV1 = /** @class */ (function (_super) {
    __extends(EventNotificationsV1, _super);
    /**
     * Construct a EventNotificationsV1 object.
     *
     * @param {Object} options - Options for the service.
     * @param {string} [options.serviceUrl] - The base url to use when contacting the service. The base url may differ between IBM Cloud regions.
     * @param {OutgoingHttpHeaders} [options.headers] - Default headers that shall be included with every request to the service.
     * @param {Authenticator} options.authenticator - The Authenticator object used to authenticate requests to the service
     * @constructor
     * @returns {EventNotificationsV1}
     */
    function EventNotificationsV1(options) {
        var _this = this;
        options = options || {};
        _this = _super.call(this, options) || this;
        if (options.serviceUrl) {
            _this.setServiceUrl(options.serviceUrl);
        }
        else {
            _this.setServiceUrl(EventNotificationsV1.DEFAULT_SERVICE_URL);
        }
        return _this;
    }
    /*************************
     * Factory method
     ************************/
    /**
     * Constructs an instance of EventNotificationsV1 with passed in options and external configuration.
     *
     * @param {UserOptions} [options] - The parameters to send to the service.
     * @param {string} [options.serviceName] - The name of the service to configure
     * @param {Authenticator} [options.authenticator] - The Authenticator object used to authenticate requests to the service
     * @param {string} [options.serviceUrl] - The URL for the service
     * @returns {EventNotificationsV1}
     */
    EventNotificationsV1.newInstance = function (options) {
        options = options || {};
        if (!options.serviceName) {
            options.serviceName = this.DEFAULT_SERVICE_NAME;
        }
        if (!options.authenticator) {
            options.authenticator = (0, ibm_cloud_sdk_core_1.getAuthenticatorFromEnvironment)(options.serviceName);
        }
        var service = new EventNotificationsV1(options);
        service.configureService(options.serviceName);
        if (options.serviceUrl) {
            service.setServiceUrl(options.serviceUrl);
        }
        return service;
    };
    /*************************
     * metrics
     ************************/
    /**
     * Get metrics.
     *
     * Get metrics.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.destinationType - Destination type. Allowed values are [smtp_custom].
     * @param {string} params.gte - GTE (greater than equal), start timestamp in UTC.
     * @param {string} params.lte - LTE (less than equal), end timestamp in UTC.
     * @param {string} [params.destinationId] - Unique identifier for Destination.
     * @param {string} [params.sourceId] - Unique identifier for Source.
     * @param {string} [params.emailTo] - Receiver email id.
     * @param {string} [params.notificationId] - Notification Id.
     * @param {string} [params.subject] - Email subject.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.Metrics>>}
     */
    EventNotificationsV1.prototype.getMetrics = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'destinationType', 'gte', 'lte'];
        var _validParams = [
            'instanceId',
            'destinationType',
            'gte',
            'lte',
            'destinationId',
            'sourceId',
            'emailTo',
            'notificationId',
            'subject',
            'headers',
        ];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var query = {
            'destination_type': _params.destinationType,
            'gte': _params.gte,
            'lte': _params.lte,
            'destination_id': _params.destinationId,
            'source_id': _params.sourceId,
            'email_to': _params.emailTo,
            'notification_id': _params.notificationId,
            'subject': _params.subject,
        };
        var path = {
            'instance_id': _params.instanceId,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'getMetrics');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/metrics',
                method: 'GET',
                qs: query,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /*************************
     * sendNotifications
     ************************/
    /**
     * Send a notification.
     *
     * Send Notifications body from the instance. For more information about Event Notifications payload, see
     * [here](https://cloud.ibm.com/docs/event-notifications?topic=event-notifications-en-spec-payload).
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {NotificationCreate} [params.body] -
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.NotificationResponse>>}
     */
    EventNotificationsV1.prototype.sendNotifications = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId'];
        var _validParams = ['instanceId', 'body', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var body = _params.body;
        var path = {
            'instance_id': _params.instanceId,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'sendNotifications');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/notifications',
                method: 'POST',
                body: body,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /*************************
     * sources
     ************************/
    /**
     * Create a new API Source.
     *
     * Create a new API Source.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.name - Name of the source.
     * @param {string} params.description - Description of the source.
     * @param {boolean} [params.enabled] - Whether the source is enabled or not.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.SourceResponse>>}
     */
    EventNotificationsV1.prototype.createSources = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'name', 'description'];
        var _validParams = ['instanceId', 'name', 'description', 'enabled', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var body = {
            'name': _params.name,
            'description': _params.description,
            'enabled': _params.enabled,
        };
        var path = {
            'instance_id': _params.instanceId,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'createSources');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/sources',
                method: 'POST',
                body: body,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * List all Sources.
     *
     * List all Sources.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {number} [params.limit] - Page limit for paginated results.
     * @param {number} [params.offset] - offset for paginated results.
     * @param {string} [params.search] - Search string for filtering results.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.SourceList>>}
     */
    EventNotificationsV1.prototype.listSources = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId'];
        var _validParams = ['instanceId', 'limit', 'offset', 'search', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var query = {
            'limit': _params.limit,
            'offset': _params.offset,
            'search': _params.search,
        };
        var path = {
            'instance_id': _params.instanceId,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'listSources');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/sources',
                method: 'GET',
                qs: query,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Get a Source.
     *
     * Get a Source.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for Source.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.Source>>}
     */
    EventNotificationsV1.prototype.getSource = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'getSource');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/sources/{id}',
                method: 'GET',
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Delete a Source.
     *
     * Delete a Source.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for Source.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.EmptyObject>>}
     */
    EventNotificationsV1.prototype.deleteSource = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'deleteSource');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/sources/{id}',
                method: 'DELETE',
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {}, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Update details of a Source.
     *
     * Update details of a Source.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for Source.
     * @param {string} [params.name] - Name of the source.
     * @param {string} [params.description] - Description of the source.
     * @param {boolean} [params.enabled] - Whether the source is enabled or not.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.Source>>}
     */
    EventNotificationsV1.prototype.updateSource = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'name', 'description', 'enabled', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var body = {
            'name': _params.name,
            'description': _params.description,
            'enabled': _params.enabled,
        };
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'updateSource');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/sources/{id}',
                method: 'PATCH',
                body: body,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /*************************
     * topics
     ************************/
    /**
     * Create a new Topic.
     *
     * Create a new Topic.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.name - Name of the topic.
     * @param {string} [params.description] - Description of the topic.
     * @param {SourcesItems[]} [params.sources] - List of sources.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.TopicResponse>>}
     */
    EventNotificationsV1.prototype.createTopic = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'name'];
        var _validParams = ['instanceId', 'name', 'description', 'sources', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var body = {
            'name': _params.name,
            'description': _params.description,
            'sources': _params.sources,
        };
        var path = {
            'instance_id': _params.instanceId,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'createTopic');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/topics',
                method: 'POST',
                body: body,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * List all Topics.
     *
     * List all Topics.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {number} [params.limit] - Page limit for paginated results.
     * @param {number} [params.offset] - offset for paginated results.
     * @param {string} [params.search] - Search string for filtering results.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.TopicList>>}
     */
    EventNotificationsV1.prototype.listTopics = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId'];
        var _validParams = ['instanceId', 'limit', 'offset', 'search', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var query = {
            'limit': _params.limit,
            'offset': _params.offset,
            'search': _params.search,
        };
        var path = {
            'instance_id': _params.instanceId,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'listTopics');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/topics',
                method: 'GET',
                qs: query,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Get details of a Topic.
     *
     * Get details of a Topic.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for Topic.
     * @param {string} [params.include] - Include sub topics.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.Topic>>}
     */
    EventNotificationsV1.prototype.getTopic = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'include', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var query = {
            'include': _params.include,
        };
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'getTopic');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/topics/{id}',
                method: 'GET',
                qs: query,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Update details of a Topic.
     *
     * Update details of a Topic.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for Topic.
     * @param {string} [params.name] - Name of the topic.
     * @param {string} [params.description] - Description of the topic.
     * @param {SourcesItems[]} [params.sources] - List of sources.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.Topic>>}
     */
    EventNotificationsV1.prototype.replaceTopic = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'name', 'description', 'sources', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var body = {
            'name': _params.name,
            'description': _params.description,
            'sources': _params.sources,
        };
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'replaceTopic');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/topics/{id}',
                method: 'PUT',
                body: body,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Delete a Topic.
     *
     * Delete a Topic.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for Topic.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.EmptyObject>>}
     */
    EventNotificationsV1.prototype.deleteTopic = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'deleteTopic');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/topics/{id}',
                method: 'DELETE',
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {}, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /*************************
     * templates
     ************************/
    /**
     * Create a new Template.
     *
     * Create a new Template.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.name - The Message Template.
     * @param {string} params.type - The type of template.
     * @param {TemplateConfigOneOf} params.params -
     * @param {string} [params.description] - The Template description.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.TemplateResponse>>}
     */
    EventNotificationsV1.prototype.createTemplate = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'name', 'type', 'params'];
        var _validParams = ['instanceId', 'name', 'type', 'params', 'description', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var body = {
            'name': _params.name,
            'type': _params.type,
            'params': _params.params,
            'description': _params.description,
        };
        var path = {
            'instance_id': _params.instanceId,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'createTemplate');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/templates',
                method: 'POST',
                body: body,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * List all templates.
     *
     * List all Templates.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {number} [params.limit] - Page limit for paginated results.
     * @param {number} [params.offset] - offset for paginated results.
     * @param {string} [params.search] - Search string for filtering results.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.TemplateList>>}
     */
    EventNotificationsV1.prototype.listTemplates = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId'];
        var _validParams = ['instanceId', 'limit', 'offset', 'search', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var query = {
            'limit': _params.limit,
            'offset': _params.offset,
            'search': _params.search,
        };
        var path = {
            'instance_id': _params.instanceId,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'listTemplates');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/templates',
                method: 'GET',
                qs: query,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Get details of a Template.
     *
     * Get details of a Template.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for Template.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.Template>>}
     */
    EventNotificationsV1.prototype.getTemplate = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'getTemplate');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/templates/{id}',
                method: 'GET',
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Update details of a Template.
     *
     * Update details of a Template.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for Template.
     * @param {string} [params.name] - Template name.
     * @param {string} [params.description] - Template description.
     * @param {string} [params.type] - The type of template.
     * @param {TemplateConfigOneOf} [params.params] -
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.Template>>}
     */
    EventNotificationsV1.prototype.replaceTemplate = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'name', 'description', 'type', 'params', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var body = {
            'name': _params.name,
            'description': _params.description,
            'type': _params.type,
            'params': _params.params,
        };
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'replaceTemplate');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/templates/{id}',
                method: 'PUT',
                body: body,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Delete a Template.
     *
     * Delete a Template.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for Template.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.EmptyObject>>}
     */
    EventNotificationsV1.prototype.deleteTemplate = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'deleteTemplate');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/templates/{id}',
                method: 'DELETE',
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {}, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * List all predefined templates.
     *
     * List all predefined templates.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.source - Source type.
     * @param {string} params.type - Destination type.
     * @param {number} [params.limit] - Page limit for paginated results.
     * @param {number} [params.offset] - offset for paginated results.
     * @param {string} [params.search] - Search string for filtering results.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.PredefinedTemplatesList>>}
     */
    EventNotificationsV1.prototype.listPreDefinedTemplates = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'source', 'type'];
        var _validParams = ['instanceId', 'source', 'type', 'limit', 'offset', 'search', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var query = {
            'source': _params.source,
            'type': _params.type,
            'limit': _params.limit,
            'offset': _params.offset,
            'search': _params.search,
        };
        var path = {
            'instance_id': _params.instanceId,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'listPreDefinedTemplates');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/pre_defined_templates',
                method: 'GET',
                qs: query,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Get details of a Predefined Template.
     *
     * Get details of a Predefined Template.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for Template.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.GetPredefinedTemplate>>}
     */
    EventNotificationsV1.prototype.getPreDefinedTemplate = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'getPreDefinedTemplate');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/pre_defined_templates/{id}',
                method: 'GET',
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /*************************
     * destinations
     ************************/
    /**
     * Create a new Destination.
     *
     * Create a new Destination.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.name - The Destination name.
     * @param {string} params.type - The type of Destination Webhook.
     * @param {string} [params.description] - The Destination description.
     * @param {boolean} [params.collectFailedEvents] - Whether to collect the failed event in Cloud Object Storage bucket.
     * @param {DestinationConfig} [params.config] - Payload describing a destination configuration.
     * @param {NodeJS.ReadableStream | Buffer} [params.certificate] - Certificate for APNS.
     * @param {string} [params.certificateContentType] - The content type of certificate.
     * @param {NodeJS.ReadableStream | Buffer} [params.icon16x16] - Safari icon 16x16.
     * @param {string} [params.icon16x16ContentType] - The content type of icon16x16.
     * @param {NodeJS.ReadableStream | Buffer} [params.icon16x162x] - Safari icon 16x16@2x.
     * @param {string} [params.icon16x162xContentType] - The content type of icon16x162x.
     * @param {NodeJS.ReadableStream | Buffer} [params.icon32x32] - Safari icon 32x32.
     * @param {string} [params.icon32x32ContentType] - The content type of icon32x32.
     * @param {NodeJS.ReadableStream | Buffer} [params.icon32x322x] - Safari icon 32x32@2x.
     * @param {string} [params.icon32x322xContentType] - The content type of icon32x322x.
     * @param {NodeJS.ReadableStream | Buffer} [params.icon128x128] - Safari icon 128x128.
     * @param {string} [params.icon128x128ContentType] - The content type of icon128x128.
     * @param {NodeJS.ReadableStream | Buffer} [params.icon128x1282x] - Safari icon 128x128@2x.
     * @param {string} [params.icon128x1282xContentType] - The content type of icon128x1282x.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.DestinationResponse>>}
     */
    EventNotificationsV1.prototype.createDestination = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'name', 'type'];
        var _validParams = [
            'instanceId',
            'name',
            'type',
            'description',
            'collectFailedEvents',
            'config',
            'certificate',
            'certificateContentType',
            'icon16x16',
            'icon16x16ContentType',
            'icon16x162x',
            'icon16x162xContentType',
            'icon32x32',
            'icon32x32ContentType',
            'icon32x322x',
            'icon32x322xContentType',
            'icon128x128',
            'icon128x128ContentType',
            'icon128x1282x',
            'icon128x1282xContentType',
            'headers',
        ];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var formData = {
            'name': _params.name,
            'type': _params.type,
            'description': _params.description,
            'collect_failed_events': _params.collectFailedEvents,
            'config': _params.config,
            'certificate': {
                data: _params.certificate,
                contentType: _params.certificateContentType,
            },
            'icon_16x16': {
                data: _params.icon16x16,
                contentType: _params.icon16x16ContentType,
            },
            'icon_16x16_2x': {
                data: _params.icon16x162x,
                contentType: _params.icon16x162xContentType,
            },
            'icon_32x32': {
                data: _params.icon32x32,
                contentType: _params.icon32x32ContentType,
            },
            'icon_32x32_2x': {
                data: _params.icon32x322x,
                contentType: _params.icon32x322xContentType,
            },
            'icon_128x128': {
                data: _params.icon128x128,
                contentType: _params.icon128x128ContentType,
            },
            'icon_128x128_2x': {
                data: _params.icon128x1282x,
                contentType: _params.icon128x1282xContentType,
            },
        };
        var path = {
            'instance_id': _params.instanceId,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'createDestination');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/destinations',
                method: 'POST',
                path: path,
                formData: formData,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                    'Content-Type': 'multipart/form-data',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * List all Destinations.
     *
     * List all Destinations.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {number} [params.limit] - Page limit for paginated results.
     * @param {number} [params.offset] - offset for paginated results.
     * @param {string} [params.search] - Search string for filtering results.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.DestinationList>>}
     */
    EventNotificationsV1.prototype.listDestinations = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId'];
        var _validParams = ['instanceId', 'limit', 'offset', 'search', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var query = {
            'limit': _params.limit,
            'offset': _params.offset,
            'search': _params.search,
        };
        var path = {
            'instance_id': _params.instanceId,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'listDestinations');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/destinations',
                method: 'GET',
                qs: query,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Get details of a Destination.
     *
     * Get details of a Destination.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for Destination.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.Destination>>}
     */
    EventNotificationsV1.prototype.getDestination = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'getDestination');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/destinations/{id}',
                method: 'GET',
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Update details of a Destination.
     *
     * Update details of a Destination.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for Destination.
     * @param {string} [params.name] - Destination name.
     * @param {string} [params.description] - Destination description.
     * @param {boolean} [params.collectFailedEvents] - Whether to collect the failed event in Cloud Object Storage bucket.
     * @param {DestinationConfig} [params.config] - Payload describing a destination configuration.
     * @param {NodeJS.ReadableStream | Buffer} [params.certificate] - Certificate for APNS.
     * @param {string} [params.certificateContentType] - The content type of certificate.
     * @param {NodeJS.ReadableStream | Buffer} [params.icon16x16] - Safari icon 16x16.
     * @param {string} [params.icon16x16ContentType] - The content type of icon16x16.
     * @param {NodeJS.ReadableStream | Buffer} [params.icon16x162x] - Safari icon 16x16@2x.
     * @param {string} [params.icon16x162xContentType] - The content type of icon16x162x.
     * @param {NodeJS.ReadableStream | Buffer} [params.icon32x32] - Safari icon 32x32.
     * @param {string} [params.icon32x32ContentType] - The content type of icon32x32.
     * @param {NodeJS.ReadableStream | Buffer} [params.icon32x322x] - Safari icon 32x32@2x.
     * @param {string} [params.icon32x322xContentType] - The content type of icon32x322x.
     * @param {NodeJS.ReadableStream | Buffer} [params.icon128x128] - Safari icon 128x128.
     * @param {string} [params.icon128x128ContentType] - The content type of icon128x128.
     * @param {NodeJS.ReadableStream | Buffer} [params.icon128x1282x] - Safari icon 128x128@2x.
     * @param {string} [params.icon128x1282xContentType] - The content type of icon128x1282x.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.Destination>>}
     */
    EventNotificationsV1.prototype.updateDestination = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = [
            'instanceId',
            'id',
            'name',
            'description',
            'collectFailedEvents',
            'config',
            'certificate',
            'certificateContentType',
            'icon16x16',
            'icon16x16ContentType',
            'icon16x162x',
            'icon16x162xContentType',
            'icon32x32',
            'icon32x32ContentType',
            'icon32x322x',
            'icon32x322xContentType',
            'icon128x128',
            'icon128x128ContentType',
            'icon128x1282x',
            'icon128x1282xContentType',
            'headers',
        ];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var formData = {
            'name': _params.name,
            'description': _params.description,
            'collect_failed_events': _params.collectFailedEvents,
            'config': _params.config,
            'certificate': {
                data: _params.certificate,
                contentType: _params.certificateContentType,
            },
            'icon_16x16': {
                data: _params.icon16x16,
                contentType: _params.icon16x16ContentType,
            },
            'icon_16x16_2x': {
                data: _params.icon16x162x,
                contentType: _params.icon16x162xContentType,
            },
            'icon_32x32': {
                data: _params.icon32x32,
                contentType: _params.icon32x32ContentType,
            },
            'icon_32x32_2x': {
                data: _params.icon32x322x,
                contentType: _params.icon32x322xContentType,
            },
            'icon_128x128': {
                data: _params.icon128x128,
                contentType: _params.icon128x128ContentType,
            },
            'icon_128x128_2x': {
                data: _params.icon128x1282x,
                contentType: _params.icon128x1282xContentType,
            },
        };
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'updateDestination');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/destinations/{id}',
                method: 'PATCH',
                path: path,
                formData: formData,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                    'Content-Type': 'multipart/form-data',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Delete a Destination.
     *
     * Delete a Destination.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for Destination.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.EmptyObject>>}
     */
    EventNotificationsV1.prototype.deleteDestination = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'deleteDestination');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/destinations/{id}',
                method: 'DELETE',
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {}, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Get enabled country details of SMS destination.
     *
     * Get enabled country details of SMS destination.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for Destination.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.EnabledCountriesResponse>>}
     */
    EventNotificationsV1.prototype.getEnabledCountries = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'getEnabledCountries');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/destinations/{id}/enabled_countries',
                method: 'GET',
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Test a Destination.
     *
     * Test a Destination.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for Destination.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.TestDestinationResponse>>}
     */
    EventNotificationsV1.prototype.testDestination = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'testDestination');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/destinations/{id}/test',
                method: 'POST',
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Verify SPF and DKIM records of custom domain.
     *
     * Verify SPF and DKIM records of custom domain.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for Destination.
     * @param {string} params.type - Verification type.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.VerificationResponse>>}
     */
    EventNotificationsV1.prototype.updateVerifyDestination = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id', 'type'];
        var _validParams = ['instanceId', 'id', 'type', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var query = {
            'type': _params.type,
        };
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'updateVerifyDestination');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/destinations/{id}/verify',
                method: 'PATCH',
                qs: query,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /*************************
     * pushDestinationAPIs
     ************************/
    /**
     * Create a new tag subscription.
     *
     * Create a new tag subscription.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for Destination.
     * @param {string} params.deviceId - Unique identifier of the device.
     * @param {string} params.tagName - The name of the tag its subscribed.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.DestinationTagsSubscriptionResponse>>}
     */
    EventNotificationsV1.prototype.createTagsSubscription = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id', 'deviceId', 'tagName'];
        var _validParams = ['instanceId', 'id', 'deviceId', 'tagName', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var body = {
            'device_id': _params.deviceId,
            'tag_name': _params.tagName,
        };
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'createTagsSubscription');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/destinations/{id}/tag_subscriptions',
                method: 'POST',
                body: body,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * List all tag subscriptions.
     *
     * List all tag subscriptions.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for Destination.
     * @param {string} [params.deviceId] - Device ID of the destination tagsubscription.
     * @param {string} [params.userId] - UserID of the destination.
     * @param {string} [params.tagName] - TagName of the subscription.
     * @param {number} [params.limit] - Page limit for paginated results.
     * @param {number} [params.offset] - offset for paginated results.
     * @param {string} [params.search] - Search string for filtering results.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.TagsSubscriptionList>>}
     */
    EventNotificationsV1.prototype.listTagsSubscription = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = [
            'instanceId',
            'id',
            'deviceId',
            'userId',
            'tagName',
            'limit',
            'offset',
            'search',
            'headers',
        ];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var query = {
            'device_id': _params.deviceId,
            'user_id': _params.userId,
            'tag_name': _params.tagName,
            'limit': _params.limit,
            'offset': _params.offset,
            'search': _params.search,
        };
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'listTagsSubscription');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/destinations/{id}/tag_subscriptions',
                method: 'GET',
                qs: query,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Delete a tag subscription.
     *
     * Delete a tag subscription.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for Destination.
     * @param {string} [params.deviceId] - Device ID of the destination tagsubscription.
     * @param {string} [params.tagName] - TagName of the subscription.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.EmptyObject>>}
     */
    EventNotificationsV1.prototype.deleteTagsSubscription = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'deviceId', 'tagName', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var query = {
            'device_id': _params.deviceId,
            'tag_name': _params.tagName,
        };
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'deleteTagsSubscription');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/destinations/{id}/tag_subscriptions',
                method: 'DELETE',
                qs: query,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {}, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /*************************
     * subscriptions
     ************************/
    /**
     * Create a new Subscription.
     *
     * Create a new Subscription.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.name - Subscription name.
     * @param {string} params.destinationId - Destination ID.
     * @param {string} params.topicId - Topic ID.
     * @param {string} [params.description] - Subscription description.
     * @param {SubscriptionCreateAttributes} [params.attributes] -
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.Subscription>>}
     */
    EventNotificationsV1.prototype.createSubscription = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'name', 'destinationId', 'topicId'];
        var _validParams = [
            'instanceId',
            'name',
            'destinationId',
            'topicId',
            'description',
            'attributes',
            'headers',
        ];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var body = {
            'name': _params.name,
            'destination_id': _params.destinationId,
            'topic_id': _params.topicId,
            'description': _params.description,
            'attributes': _params.attributes,
        };
        var path = {
            'instance_id': _params.instanceId,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'createSubscription');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/subscriptions',
                method: 'POST',
                body: body,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * List all Subscriptions.
     *
     * List all Subscriptions.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {number} [params.offset] - offset for paginated results.
     * @param {number} [params.limit] - Page limit for paginated results.
     * @param {string} [params.search] - Search string for filtering results.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.SubscriptionList>>}
     */
    EventNotificationsV1.prototype.listSubscriptions = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId'];
        var _validParams = ['instanceId', 'offset', 'limit', 'search', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var query = {
            'offset': _params.offset,
            'limit': _params.limit,
            'search': _params.search,
        };
        var path = {
            'instance_id': _params.instanceId,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'listSubscriptions');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/subscriptions',
                method: 'GET',
                qs: query,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Get details of a Subscription.
     *
     * Get details of a Subscription.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for Subscription.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.Subscription>>}
     */
    EventNotificationsV1.prototype.getSubscription = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'getSubscription');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/subscriptions/{id}',
                method: 'GET',
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Delete a Subscription.
     *
     * Delete a Subscription.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for Subscription.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.EmptyObject>>}
     */
    EventNotificationsV1.prototype.deleteSubscription = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'deleteSubscription');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/subscriptions/{id}',
                method: 'DELETE',
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {}, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Update details of a Subscription.
     *
     * Update details of a Subscription.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for Subscription.
     * @param {string} [params.name] - Name of the subscription.
     * @param {string} [params.description] - Description of the subscription.
     * @param {SubscriptionUpdateAttributes} [params.attributes] -
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.Subscription>>}
     */
    EventNotificationsV1.prototype.updateSubscription = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'name', 'description', 'attributes', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var body = {
            'name': _params.name,
            'description': _params.description,
            'attributes': _params.attributes,
        };
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'updateSubscription');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/subscriptions/{id}',
                method: 'PATCH',
                body: body,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /*************************
     * integrations
     ************************/
    /**
     * Create an Integration.
     *
     * Create an Integration.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.type - The type of Integration.
     * @param {IntegrationCreateMetadata} params.metadata - Integration Metadata object.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.IntegrationCreateResponse>>}
     */
    EventNotificationsV1.prototype.createIntegration = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'type', 'metadata'];
        var _validParams = ['instanceId', 'type', 'metadata', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var body = {
            'type': _params.type,
            'metadata': _params.metadata,
        };
        var path = {
            'instance_id': _params.instanceId,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'createIntegration');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/integrations',
                method: 'POST',
                body: body,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * List all Integrations.
     *
     * List of all KMS Integrations.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {number} [params.offset] - offset for paginated results.
     * @param {number} [params.limit] - Page limit for paginated results.
     * @param {string} [params.search] - Search string for filtering results.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.IntegrationList>>}
     */
    EventNotificationsV1.prototype.listIntegrations = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId'];
        var _validParams = ['instanceId', 'offset', 'limit', 'search', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var query = {
            'offset': _params.offset,
            'limit': _params.limit,
            'search': _params.search,
        };
        var path = {
            'instance_id': _params.instanceId,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'listIntegrations');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/integrations',
                method: 'GET',
                qs: query,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Get a single Integration.
     *
     * Get a single KMS Integration.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for integration.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.IntegrationGetResponse>>}
     */
    EventNotificationsV1.prototype.getIntegration = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'getIntegration');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/integrations/{id}',
                method: 'GET',
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Update an existing Integration.
     *
     * Update an existing Integration.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for integration.
     * @param {string} params.type - Integration type. Allowed values are kms, hs-crypto and collect_failed_events.
     * @param {IntegrationMetadata} params.metadata - Integration Metadata object.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.IntegrationGetResponse>>}
     */
    EventNotificationsV1.prototype.replaceIntegration = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id', 'type', 'metadata'];
        var _validParams = ['instanceId', 'id', 'type', 'metadata', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var body = {
            'type': _params.type,
            'metadata': _params.metadata,
        };
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'replaceIntegration');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/integrations/{id}',
                method: 'PUT',
                body: body,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /*************************
     * sMTPConfigurations
     ************************/
    /**
     * Create a new SMTP Configuration.
     *
     * Create a new SMTP Configuration.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.name - The name of SMTP configuration.
     * @param {string} params.domain - Domain Name.
     * @param {string} [params.description] - The description of SMTP configuration.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.SMTPCreateResponse>>}
     */
    EventNotificationsV1.prototype.createSmtpConfiguration = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'name', 'domain'];
        var _validParams = ['instanceId', 'name', 'domain', 'description', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var body = {
            'name': _params.name,
            'domain': _params.domain,
            'description': _params.description,
        };
        var path = {
            'instance_id': _params.instanceId,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'createSmtpConfiguration');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/smtp/config',
                method: 'POST',
                body: body,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * List all SMTP Configurations.
     *
     * List all SMTP Configurations.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {number} [params.limit] - Page limit for paginated results.
     * @param {number} [params.offset] - offset for paginated results.
     * @param {string} [params.search] - Search string for filtering results.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.SMTPConfigurationsList>>}
     */
    EventNotificationsV1.prototype.listSmtpConfigurations = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId'];
        var _validParams = ['instanceId', 'limit', 'offset', 'search', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var query = {
            'limit': _params.limit,
            'offset': _params.offset,
            'search': _params.search,
        };
        var path = {
            'instance_id': _params.instanceId,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'listSmtpConfigurations');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/smtp/config',
                method: 'GET',
                qs: query,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Create a new SMTP User.
     *
     * Create a new SMTP User.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for SMTP.
     * @param {string} [params.description] - The description of SMTP configuration.
     * @param {string} [params.usernameToClone] - provide name of the user to clone.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.SMTPUserResponse>>}
     */
    EventNotificationsV1.prototype.createSmtpUser = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'description', 'usernameToClone', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var body = {
            'description': _params.description,
        };
        var query = {
            'username_to_clone': _params.usernameToClone,
        };
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'createSmtpUser');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/smtp/config/{id}/users',
                method: 'POST',
                body: body,
                qs: query,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * List all SMTP users.
     *
     * List all SMTP users.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for SMTP.
     * @param {number} [params.limit] - Page limit for paginated results.
     * @param {number} [params.offset] - offset for paginated results.
     * @param {string} [params.search] - Search string for filtering results.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.SMTPUsersList>>}
     */
    EventNotificationsV1.prototype.listSmtpUsers = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'limit', 'offset', 'search', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var query = {
            'limit': _params.limit,
            'offset': _params.offset,
            'search': _params.search,
        };
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'listSmtpUsers');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/smtp/config/{id}/users',
                method: 'GET',
                qs: query,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Get details of a SMTP Configuration.
     *
     * Get details of a SMTP Configuration.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for SMTP.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.SMTPConfiguration>>}
     */
    EventNotificationsV1.prototype.getSmtpConfiguration = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'getSmtpConfiguration');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/smtp/config/{id}',
                method: 'GET',
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Update details of SMTP Configuration.
     *
     * Update details of SMTP Configuration.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for SMTP.
     * @param {string} [params.name] - SMTP name.
     * @param {string} [params.description] - SMTP description.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.SMTPConfiguration>>}
     */
    EventNotificationsV1.prototype.updateSmtpConfiguration = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'name', 'description', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var body = {
            'name': _params.name,
            'description': _params.description,
        };
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'updateSmtpConfiguration');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/smtp/config/{id}',
                method: 'PATCH',
                body: body,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Delete a SMTP Configuration.
     *
     * Delete a SMTP Configuration.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for SMTP.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.EmptyObject>>}
     */
    EventNotificationsV1.prototype.deleteSmtpConfiguration = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'deleteSmtpConfiguration');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/smtp/config/{id}',
                method: 'DELETE',
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {}, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Get details of a SMTP User.
     *
     * Get details of a SMTP User.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for SMTP.
     * @param {string} params.userId - UserID.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.SMTPUser>>}
     */
    EventNotificationsV1.prototype.getSmtpUser = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id', 'userId'];
        var _validParams = ['instanceId', 'id', 'userId', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
            'user_id': _params.userId,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'getSmtpUser');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/smtp/config/{id}/users/{user_id}',
                method: 'GET',
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Update details of a SMTP User.
     *
     * Update details of a SMTP User.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for SMTP.
     * @param {string} params.userId - UserID.
     * @param {string} [params.description] - SMTP user description.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.SMTPUser>>}
     */
    EventNotificationsV1.prototype.updateSmtpUser = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id', 'userId'];
        var _validParams = ['instanceId', 'id', 'userId', 'description', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var body = {
            'description': _params.description,
        };
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
            'user_id': _params.userId,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'updateSmtpUser');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/smtp/config/{id}/users/{user_id}',
                method: 'PATCH',
                body: body,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Delete a SMTP user.
     *
     * Delete a SMTP user.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for SMTP.
     * @param {string} params.userId - UserID.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.EmptyObject>>}
     */
    EventNotificationsV1.prototype.deleteSmtpUser = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id', 'userId'];
        var _validParams = ['instanceId', 'id', 'userId', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
            'user_id': _params.userId,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'deleteSmtpUser');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/smtp/config/{id}/users/{user_id}',
                method: 'DELETE',
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {}, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Get details of SMTP configuration allowed IPs.
     *
     * Get details of SMTP configuration allowed IPs.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for SMTP.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.SMTPAllowedIPs>>}
     */
    EventNotificationsV1.prototype.getSmtpAllowedIps = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'getSmtpAllowedIps');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/smtp/config/{id}/allowed_ips',
                method: 'GET',
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /**
     * Verify SMTP configuration domain.
     *
     * Verify SMTP configuration domain.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Unique identifier for SMTP.
     * @param {string} params.type - SMTP verification type.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.SMTPVerificationUpdateResponse>>}
     */
    EventNotificationsV1.prototype.updateVerifySmtp = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id', 'type'];
        var _validParams = ['instanceId', 'id', 'type', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var query = {
            'type': _params.type,
        };
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'updateVerifySmtp');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/smtp/config/{id}/verify',
                method: 'PATCH',
                qs: query,
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    /*************************
     * getNotificationsStatus
     ************************/
    /**
     * Get notification status.
     *
     * Get notification status.
     *
     * @param {Object} params - The parameters to send to the service.
     * @param {string} params.instanceId - Unique identifier for IBM Cloud Event Notifications instance.
     * @param {string} params.id - Get Notification Id.
     * @param {OutgoingHttpHeaders} [params.headers] - Custom request headers
     * @returns {Promise<EventNotificationsV1.Response<EventNotificationsV1.GetNotificationStatusResponse>>}
     */
    EventNotificationsV1.prototype.getNotificationsStatus = function (params) {
        var _params = __assign({}, params);
        var _requiredParams = ['instanceId', 'id'];
        var _validParams = ['instanceId', 'id', 'headers'];
        var _validationErrors = (0, ibm_cloud_sdk_core_1.validateParams)(_params, _requiredParams, _validParams);
        if (_validationErrors) {
            return Promise.reject(_validationErrors);
        }
        var path = {
            'instance_id': _params.instanceId,
            'id': _params.id,
        };
        var sdkHeaders = (0, common_1.getSdkHeaders)(EventNotificationsV1.DEFAULT_SERVICE_NAME, 'v1', 'getNotificationsStatus');
        var parameters = {
            options: {
                url: '/v1/instances/{instance_id}/notifications/{id}',
                method: 'GET',
                path: path,
            },
            defaultOptions: extend(true, {}, this.baseOptions, {
                headers: extend(true, sdkHeaders, {
                    'Accept': 'application/json',
                }, _params.headers),
            }),
        };
        return this.createRequest(parameters);
    };
    EventNotificationsV1.DEFAULT_SERVICE_URL = 'https://us-south.event-notifications.cloud.ibm.com/event-notifications';
    EventNotificationsV1.DEFAULT_SERVICE_NAME = 'event_notifications';
    return EventNotificationsV1;
}(ibm_cloud_sdk_core_1.BaseService));
/*************************
 * interfaces
 ************************/
(function (EventNotificationsV1) {
    /** Constants for the `getMetrics` operation. */
    var GetMetricsConstants;
    (function (GetMetricsConstants) {
        /** Destination type. Allowed values are [smtp_custom]. */
        var DestinationType;
        (function (DestinationType) {
            DestinationType["SMTP_CUSTOM"] = "smtp_custom";
        })(DestinationType = GetMetricsConstants.DestinationType || (GetMetricsConstants.DestinationType = {}));
    })(GetMetricsConstants = EventNotificationsV1.GetMetricsConstants || (EventNotificationsV1.GetMetricsConstants = {}));
    /** Constants for the `createDestination` operation. */
    var CreateDestinationConstants;
    (function (CreateDestinationConstants) {
        /** The type of Destination Webhook. */
        var Type;
        (function (Type) {
            Type["WEBHOOK"] = "webhook";
            Type["PUSH_ANDROID"] = "push_android";
            Type["PUSH_IOS"] = "push_ios";
            Type["PUSH_CHROME"] = "push_chrome";
            Type["PUSH_FIREFOX"] = "push_firefox";
            Type["SLACK"] = "slack";
            Type["IBMCE"] = "ibmce";
            Type["PAGERDUTY"] = "pagerduty";
            Type["PUSH_SAFARI"] = "push_safari";
            Type["MSTEAMS"] = "msteams";
            Type["SERVICENOW"] = "servicenow";
            Type["IBMCOS"] = "ibmcos";
            Type["PUSH_HUAWEI"] = "push_huawei";
            Type["SMTP_CUSTOM"] = "smtp_custom";
            Type["SMS_CUSTOM"] = "sms_custom";
            Type["EVENT_STREAMS"] = "event_streams";
            Type["APP_CONFIGURATION"] = "app_configuration";
        })(Type = CreateDestinationConstants.Type || (CreateDestinationConstants.Type = {}));
    })(CreateDestinationConstants = EventNotificationsV1.CreateDestinationConstants || (EventNotificationsV1.CreateDestinationConstants = {}));
    /** Constants for the `createIntegration` operation. */
    var CreateIntegrationConstants;
    (function (CreateIntegrationConstants) {
        /** The type of Integration. */
        var Type;
        (function (Type) {
            Type["COLLECT_FAILED_EVENTS"] = "collect_failed_events";
        })(Type = CreateIntegrationConstants.Type || (CreateIntegrationConstants.Type = {}));
    })(CreateIntegrationConstants = EventNotificationsV1.CreateIntegrationConstants || (EventNotificationsV1.CreateIntegrationConstants = {}));
    /*************************
     * pager classes
     ************************/
    /**
     * SourcesPager can be used to simplify the use of listSources().
     */
    var SourcesPager = /** @class */ (function () {
        /**
         * Construct a SourcesPager object.
         *
         * @param {EventNotificationsV1}  client - The service client instance used to invoke listSources()
         * @param {Object} params - The parameters to be passed to listSources()
         * @constructor
         * @returns {SourcesPager}
         */
        function SourcesPager(client, params) {
            if (params && params.offset) {
                throw new Error("the params.offset field should not be set");
            }
            this._hasNext = true;
            this.pageContext = { next: undefined };
            this.client = client;
            this.params = JSON.parse(JSON.stringify(params || {}));
        }
        /**
         * Returns true if there are potentially more results to be retrieved by invoking getNext().
         * @returns {boolean}
         */
        SourcesPager.prototype.hasNext = function () {
            return this._hasNext;
        };
        /**
         * Returns the next page of results by invoking listSources().
         * @returns {Promise<EventNotificationsV1.SourceListItem[]>}
         */
        SourcesPager.prototype.getNext = function () {
            return __awaiter(this, void 0, void 0, function () {
                var response, result, next;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!this.hasNext()) {
                                throw new Error('No more results available');
                            }
                            if (this.pageContext.next) {
                                this.params.offset = this.pageContext.next;
                            }
                            return [4 /*yield*/, this.client.listSources(this.params)];
                        case 1:
                            response = _a.sent();
                            result = response.result;
                            next = null;
                            if (result && result.next) {
                                if (result.next.href) {
                                    next = (0, ibm_cloud_sdk_core_2.getQueryParam)(result.next.href, 'offset');
                                }
                            }
                            this.pageContext.next = next;
                            if (!this.pageContext.next) {
                                this._hasNext = false;
                            }
                            return [2 /*return*/, result.sources];
                    }
                });
            });
        };
        /**
         * Returns all results by invoking listSources() repeatedly until all pages of results have been retrieved.
         * @returns {Promise<EventNotificationsV1.SourceListItem[]>}
         */
        SourcesPager.prototype.getAll = function () {
            return __awaiter(this, void 0, void 0, function () {
                var results, nextPage;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            results = [];
                            _a.label = 1;
                        case 1:
                            if (!this.hasNext()) return [3 /*break*/, 3];
                            return [4 /*yield*/, this.getNext()];
                        case 2:
                            nextPage = _a.sent();
                            results.push.apply(results, nextPage);
                            return [3 /*break*/, 1];
                        case 3: return [2 /*return*/, results];
                    }
                });
            });
        };
        return SourcesPager;
    }());
    EventNotificationsV1.SourcesPager = SourcesPager;
    /**
     * TopicsPager can be used to simplify the use of listTopics().
     */
    var TopicsPager = /** @class */ (function () {
        /**
         * Construct a TopicsPager object.
         *
         * @param {EventNotificationsV1}  client - The service client instance used to invoke listTopics()
         * @param {Object} params - The parameters to be passed to listTopics()
         * @constructor
         * @returns {TopicsPager}
         */
        function TopicsPager(client, params) {
            if (params && params.offset) {
                throw new Error("the params.offset field should not be set");
            }
            this._hasNext = true;
            this.pageContext = { next: undefined };
            this.client = client;
            this.params = JSON.parse(JSON.stringify(params || {}));
        }
        /**
         * Returns true if there are potentially more results to be retrieved by invoking getNext().
         * @returns {boolean}
         */
        TopicsPager.prototype.hasNext = function () {
            return this._hasNext;
        };
        /**
         * Returns the next page of results by invoking listTopics().
         * @returns {Promise<EventNotificationsV1.TopicsListItem[]>}
         */
        TopicsPager.prototype.getNext = function () {
            return __awaiter(this, void 0, void 0, function () {
                var response, result, next;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!this.hasNext()) {
                                throw new Error('No more results available');
                            }
                            if (this.pageContext.next) {
                                this.params.offset = this.pageContext.next;
                            }
                            return [4 /*yield*/, this.client.listTopics(this.params)];
                        case 1:
                            response = _a.sent();
                            result = response.result;
                            next = null;
                            if (result && result.next) {
                                if (result.next.href) {
                                    next = (0, ibm_cloud_sdk_core_2.getQueryParam)(result.next.href, 'offset');
                                }
                            }
                            this.pageContext.next = next;
                            if (!this.pageContext.next) {
                                this._hasNext = false;
                            }
                            return [2 /*return*/, result.topics];
                    }
                });
            });
        };
        /**
         * Returns all results by invoking listTopics() repeatedly until all pages of results have been retrieved.
         * @returns {Promise<EventNotificationsV1.TopicsListItem[]>}
         */
        TopicsPager.prototype.getAll = function () {
            return __awaiter(this, void 0, void 0, function () {
                var results, nextPage;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            results = [];
                            _a.label = 1;
                        case 1:
                            if (!this.hasNext()) return [3 /*break*/, 3];
                            return [4 /*yield*/, this.getNext()];
                        case 2:
                            nextPage = _a.sent();
                            results.push.apply(results, nextPage);
                            return [3 /*break*/, 1];
                        case 3: return [2 /*return*/, results];
                    }
                });
            });
        };
        return TopicsPager;
    }());
    EventNotificationsV1.TopicsPager = TopicsPager;
    /**
     * TemplatesPager can be used to simplify the use of listTemplates().
     */
    var TemplatesPager = /** @class */ (function () {
        /**
         * Construct a TemplatesPager object.
         *
         * @param {EventNotificationsV1}  client - The service client instance used to invoke listTemplates()
         * @param {Object} params - The parameters to be passed to listTemplates()
         * @constructor
         * @returns {TemplatesPager}
         */
        function TemplatesPager(client, params) {
            if (params && params.offset) {
                throw new Error("the params.offset field should not be set");
            }
            this._hasNext = true;
            this.pageContext = { next: undefined };
            this.client = client;
            this.params = JSON.parse(JSON.stringify(params || {}));
        }
        /**
         * Returns true if there are potentially more results to be retrieved by invoking getNext().
         * @returns {boolean}
         */
        TemplatesPager.prototype.hasNext = function () {
            return this._hasNext;
        };
        /**
         * Returns the next page of results by invoking listTemplates().
         * @returns {Promise<EventNotificationsV1.Template[]>}
         */
        TemplatesPager.prototype.getNext = function () {
            return __awaiter(this, void 0, void 0, function () {
                var response, result, next;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!this.hasNext()) {
                                throw new Error('No more results available');
                            }
                            if (this.pageContext.next) {
                                this.params.offset = this.pageContext.next;
                            }
                            return [4 /*yield*/, this.client.listTemplates(this.params)];
                        case 1:
                            response = _a.sent();
                            result = response.result;
                            next = null;
                            if (result && result.next) {
                                if (result.next.href) {
                                    next = (0, ibm_cloud_sdk_core_2.getQueryParam)(result.next.href, 'offset');
                                }
                            }
                            this.pageContext.next = next;
                            if (!this.pageContext.next) {
                                this._hasNext = false;
                            }
                            return [2 /*return*/, result.templates];
                    }
                });
            });
        };
        /**
         * Returns all results by invoking listTemplates() repeatedly until all pages of results have been retrieved.
         * @returns {Promise<EventNotificationsV1.Template[]>}
         */
        TemplatesPager.prototype.getAll = function () {
            return __awaiter(this, void 0, void 0, function () {
                var results, nextPage;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            results = [];
                            _a.label = 1;
                        case 1:
                            if (!this.hasNext()) return [3 /*break*/, 3];
                            return [4 /*yield*/, this.getNext()];
                        case 2:
                            nextPage = _a.sent();
                            results.push.apply(results, nextPage);
                            return [3 /*break*/, 1];
                        case 3: return [2 /*return*/, results];
                    }
                });
            });
        };
        return TemplatesPager;
    }());
    EventNotificationsV1.TemplatesPager = TemplatesPager;
    /**
     * PreDefinedTemplatesPager can be used to simplify the use of listPreDefinedTemplates().
     */
    var PreDefinedTemplatesPager = /** @class */ (function () {
        /**
         * Construct a PreDefinedTemplatesPager object.
         *
         * @param {EventNotificationsV1}  client - The service client instance used to invoke listPreDefinedTemplates()
         * @param {Object} params - The parameters to be passed to listPreDefinedTemplates()
         * @constructor
         * @returns {PreDefinedTemplatesPager}
         */
        function PreDefinedTemplatesPager(client, params) {
            if (params && params.offset) {
                throw new Error("the params.offset field should not be set");
            }
            this._hasNext = true;
            this.pageContext = { next: undefined };
            this.client = client;
            this.params = JSON.parse(JSON.stringify(params || {}));
        }
        /**
         * Returns true if there are potentially more results to be retrieved by invoking getNext().
         * @returns {boolean}
         */
        PreDefinedTemplatesPager.prototype.hasNext = function () {
            return this._hasNext;
        };
        /**
         * Returns the next page of results by invoking listPreDefinedTemplates().
         * @returns {Promise<EventNotificationsV1.PredefinedTemplate[]>}
         */
        PreDefinedTemplatesPager.prototype.getNext = function () {
            return __awaiter(this, void 0, void 0, function () {
                var response, result, next;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!this.hasNext()) {
                                throw new Error('No more results available');
                            }
                            if (this.pageContext.next) {
                                this.params.offset = this.pageContext.next;
                            }
                            return [4 /*yield*/, this.client.listPreDefinedTemplates(this.params)];
                        case 1:
                            response = _a.sent();
                            result = response.result;
                            next = null;
                            if (result && result.next) {
                                if (result.next.href) {
                                    next = (0, ibm_cloud_sdk_core_2.getQueryParam)(result.next.href, 'offset');
                                }
                            }
                            this.pageContext.next = next;
                            if (!this.pageContext.next) {
                                this._hasNext = false;
                            }
                            return [2 /*return*/, result.templates];
                    }
                });
            });
        };
        /**
         * Returns all results by invoking listPreDefinedTemplates() repeatedly until all pages of results have been retrieved.
         * @returns {Promise<EventNotificationsV1.PredefinedTemplate[]>}
         */
        PreDefinedTemplatesPager.prototype.getAll = function () {
            return __awaiter(this, void 0, void 0, function () {
                var results, nextPage;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            results = [];
                            _a.label = 1;
                        case 1:
                            if (!this.hasNext()) return [3 /*break*/, 3];
                            return [4 /*yield*/, this.getNext()];
                        case 2:
                            nextPage = _a.sent();
                            results.push.apply(results, nextPage);
                            return [3 /*break*/, 1];
                        case 3: return [2 /*return*/, results];
                    }
                });
            });
        };
        return PreDefinedTemplatesPager;
    }());
    EventNotificationsV1.PreDefinedTemplatesPager = PreDefinedTemplatesPager;
    /**
     * DestinationsPager can be used to simplify the use of listDestinations().
     */
    var DestinationsPager = /** @class */ (function () {
        /**
         * Construct a DestinationsPager object.
         *
         * @param {EventNotificationsV1}  client - The service client instance used to invoke listDestinations()
         * @param {Object} params - The parameters to be passed to listDestinations()
         * @constructor
         * @returns {DestinationsPager}
         */
        function DestinationsPager(client, params) {
            if (params && params.offset) {
                throw new Error("the params.offset field should not be set");
            }
            this._hasNext = true;
            this.pageContext = { next: undefined };
            this.client = client;
            this.params = JSON.parse(JSON.stringify(params || {}));
        }
        /**
         * Returns true if there are potentially more results to be retrieved by invoking getNext().
         * @returns {boolean}
         */
        DestinationsPager.prototype.hasNext = function () {
            return this._hasNext;
        };
        /**
         * Returns the next page of results by invoking listDestinations().
         * @returns {Promise<EventNotificationsV1.DestinationListItem[]>}
         */
        DestinationsPager.prototype.getNext = function () {
            return __awaiter(this, void 0, void 0, function () {
                var response, result, next;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!this.hasNext()) {
                                throw new Error('No more results available');
                            }
                            if (this.pageContext.next) {
                                this.params.offset = this.pageContext.next;
                            }
                            return [4 /*yield*/, this.client.listDestinations(this.params)];
                        case 1:
                            response = _a.sent();
                            result = response.result;
                            next = null;
                            if (result && result.next) {
                                if (result.next.href) {
                                    next = (0, ibm_cloud_sdk_core_2.getQueryParam)(result.next.href, 'offset');
                                }
                            }
                            this.pageContext.next = next;
                            if (!this.pageContext.next) {
                                this._hasNext = false;
                            }
                            return [2 /*return*/, result.destinations];
                    }
                });
            });
        };
        /**
         * Returns all results by invoking listDestinations() repeatedly until all pages of results have been retrieved.
         * @returns {Promise<EventNotificationsV1.DestinationListItem[]>}
         */
        DestinationsPager.prototype.getAll = function () {
            return __awaiter(this, void 0, void 0, function () {
                var results, nextPage;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            results = [];
                            _a.label = 1;
                        case 1:
                            if (!this.hasNext()) return [3 /*break*/, 3];
                            return [4 /*yield*/, this.getNext()];
                        case 2:
                            nextPage = _a.sent();
                            results.push.apply(results, nextPage);
                            return [3 /*break*/, 1];
                        case 3: return [2 /*return*/, results];
                    }
                });
            });
        };
        return DestinationsPager;
    }());
    EventNotificationsV1.DestinationsPager = DestinationsPager;
    /**
     * TagsSubscriptionPager can be used to simplify the use of listTagsSubscription().
     */
    var TagsSubscriptionPager = /** @class */ (function () {
        /**
         * Construct a TagsSubscriptionPager object.
         *
         * @param {EventNotificationsV1}  client - The service client instance used to invoke listTagsSubscription()
         * @param {Object} params - The parameters to be passed to listTagsSubscription()
         * @constructor
         * @returns {TagsSubscriptionPager}
         */
        function TagsSubscriptionPager(client, params) {
            if (params && params.offset) {
                throw new Error("the params.offset field should not be set");
            }
            this._hasNext = true;
            this.pageContext = { next: undefined };
            this.client = client;
            this.params = JSON.parse(JSON.stringify(params || {}));
        }
        /**
         * Returns true if there are potentially more results to be retrieved by invoking getNext().
         * @returns {boolean}
         */
        TagsSubscriptionPager.prototype.hasNext = function () {
            return this._hasNext;
        };
        /**
         * Returns the next page of results by invoking listTagsSubscription().
         * @returns {Promise<EventNotificationsV1.TagsSubscriptionListItem[]>}
         */
        TagsSubscriptionPager.prototype.getNext = function () {
            return __awaiter(this, void 0, void 0, function () {
                var response, result, next;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!this.hasNext()) {
                                throw new Error('No more results available');
                            }
                            if (this.pageContext.next) {
                                this.params.offset = this.pageContext.next;
                            }
                            return [4 /*yield*/, this.client.listTagsSubscription(this.params)];
                        case 1:
                            response = _a.sent();
                            result = response.result;
                            next = null;
                            if (result && result.next) {
                                if (result.next.href) {
                                    next = (0, ibm_cloud_sdk_core_2.getQueryParam)(result.next.href, 'offset');
                                }
                            }
                            this.pageContext.next = next;
                            if (!this.pageContext.next) {
                                this._hasNext = false;
                            }
                            return [2 /*return*/, result.tag_subscriptions];
                    }
                });
            });
        };
        /**
         * Returns all results by invoking listTagsSubscription() repeatedly until all pages of results have been retrieved.
         * @returns {Promise<EventNotificationsV1.TagsSubscriptionListItem[]>}
         */
        TagsSubscriptionPager.prototype.getAll = function () {
            return __awaiter(this, void 0, void 0, function () {
                var results, nextPage;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            results = [];
                            _a.label = 1;
                        case 1:
                            if (!this.hasNext()) return [3 /*break*/, 3];
                            return [4 /*yield*/, this.getNext()];
                        case 2:
                            nextPage = _a.sent();
                            results.push.apply(results, nextPage);
                            return [3 /*break*/, 1];
                        case 3: return [2 /*return*/, results];
                    }
                });
            });
        };
        return TagsSubscriptionPager;
    }());
    EventNotificationsV1.TagsSubscriptionPager = TagsSubscriptionPager;
    /**
     * SubscriptionsPager can be used to simplify the use of listSubscriptions().
     */
    var SubscriptionsPager = /** @class */ (function () {
        /**
         * Construct a SubscriptionsPager object.
         *
         * @param {EventNotificationsV1}  client - The service client instance used to invoke listSubscriptions()
         * @param {Object} params - The parameters to be passed to listSubscriptions()
         * @constructor
         * @returns {SubscriptionsPager}
         */
        function SubscriptionsPager(client, params) {
            if (params && params.offset) {
                throw new Error("the params.offset field should not be set");
            }
            this._hasNext = true;
            this.pageContext = { next: undefined };
            this.client = client;
            this.params = JSON.parse(JSON.stringify(params || {}));
        }
        /**
         * Returns true if there are potentially more results to be retrieved by invoking getNext().
         * @returns {boolean}
         */
        SubscriptionsPager.prototype.hasNext = function () {
            return this._hasNext;
        };
        /**
         * Returns the next page of results by invoking listSubscriptions().
         * @returns {Promise<EventNotificationsV1.SubscriptionListItem[]>}
         */
        SubscriptionsPager.prototype.getNext = function () {
            return __awaiter(this, void 0, void 0, function () {
                var response, result, next;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!this.hasNext()) {
                                throw new Error('No more results available');
                            }
                            if (this.pageContext.next) {
                                this.params.offset = this.pageContext.next;
                            }
                            return [4 /*yield*/, this.client.listSubscriptions(this.params)];
                        case 1:
                            response = _a.sent();
                            result = response.result;
                            next = null;
                            if (result && result.next) {
                                if (result.next.href) {
                                    next = (0, ibm_cloud_sdk_core_2.getQueryParam)(result.next.href, 'offset');
                                }
                            }
                            this.pageContext.next = next;
                            if (!this.pageContext.next) {
                                this._hasNext = false;
                            }
                            return [2 /*return*/, result.subscriptions];
                    }
                });
            });
        };
        /**
         * Returns all results by invoking listSubscriptions() repeatedly until all pages of results have been retrieved.
         * @returns {Promise<EventNotificationsV1.SubscriptionListItem[]>}
         */
        SubscriptionsPager.prototype.getAll = function () {
            return __awaiter(this, void 0, void 0, function () {
                var results, nextPage;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            results = [];
                            _a.label = 1;
                        case 1:
                            if (!this.hasNext()) return [3 /*break*/, 3];
                            return [4 /*yield*/, this.getNext()];
                        case 2:
                            nextPage = _a.sent();
                            results.push.apply(results, nextPage);
                            return [3 /*break*/, 1];
                        case 3: return [2 /*return*/, results];
                    }
                });
            });
        };
        return SubscriptionsPager;
    }());
    EventNotificationsV1.SubscriptionsPager = SubscriptionsPager;
    /**
     * IntegrationsPager can be used to simplify the use of listIntegrations().
     */
    var IntegrationsPager = /** @class */ (function () {
        /**
         * Construct a IntegrationsPager object.
         *
         * @param {EventNotificationsV1}  client - The service client instance used to invoke listIntegrations()
         * @param {Object} params - The parameters to be passed to listIntegrations()
         * @constructor
         * @returns {IntegrationsPager}
         */
        function IntegrationsPager(client, params) {
            if (params && params.offset) {
                throw new Error("the params.offset field should not be set");
            }
            this._hasNext = true;
            this.pageContext = { next: undefined };
            this.client = client;
            this.params = JSON.parse(JSON.stringify(params || {}));
        }
        /**
         * Returns true if there are potentially more results to be retrieved by invoking getNext().
         * @returns {boolean}
         */
        IntegrationsPager.prototype.hasNext = function () {
            return this._hasNext;
        };
        /**
         * Returns the next page of results by invoking listIntegrations().
         * @returns {Promise<EventNotificationsV1.IntegrationListItem[]>}
         */
        IntegrationsPager.prototype.getNext = function () {
            return __awaiter(this, void 0, void 0, function () {
                var response, result, next;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!this.hasNext()) {
                                throw new Error('No more results available');
                            }
                            if (this.pageContext.next) {
                                this.params.offset = this.pageContext.next;
                            }
                            return [4 /*yield*/, this.client.listIntegrations(this.params)];
                        case 1:
                            response = _a.sent();
                            result = response.result;
                            next = null;
                            if (result && result.next) {
                                if (result.next.href) {
                                    next = (0, ibm_cloud_sdk_core_2.getQueryParam)(result.next.href, 'offset');
                                }
                            }
                            this.pageContext.next = next;
                            if (!this.pageContext.next) {
                                this._hasNext = false;
                            }
                            return [2 /*return*/, result.integrations];
                    }
                });
            });
        };
        /**
         * Returns all results by invoking listIntegrations() repeatedly until all pages of results have been retrieved.
         * @returns {Promise<EventNotificationsV1.IntegrationListItem[]>}
         */
        IntegrationsPager.prototype.getAll = function () {
            return __awaiter(this, void 0, void 0, function () {
                var results, nextPage;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            results = [];
                            _a.label = 1;
                        case 1:
                            if (!this.hasNext()) return [3 /*break*/, 3];
                            return [4 /*yield*/, this.getNext()];
                        case 2:
                            nextPage = _a.sent();
                            results.push.apply(results, nextPage);
                            return [3 /*break*/, 1];
                        case 3: return [2 /*return*/, results];
                    }
                });
            });
        };
        return IntegrationsPager;
    }());
    EventNotificationsV1.IntegrationsPager = IntegrationsPager;
    /**
     * SmtpConfigurationsPager can be used to simplify the use of listSmtpConfigurations().
     */
    var SmtpConfigurationsPager = /** @class */ (function () {
        /**
         * Construct a SmtpConfigurationsPager object.
         *
         * @param {EventNotificationsV1}  client - The service client instance used to invoke listSmtpConfigurations()
         * @param {Object} params - The parameters to be passed to listSmtpConfigurations()
         * @constructor
         * @returns {SmtpConfigurationsPager}
         */
        function SmtpConfigurationsPager(client, params) {
            if (params && params.offset) {
                throw new Error("the params.offset field should not be set");
            }
            this._hasNext = true;
            this.pageContext = { next: undefined };
            this.client = client;
            this.params = JSON.parse(JSON.stringify(params || {}));
        }
        /**
         * Returns true if there are potentially more results to be retrieved by invoking getNext().
         * @returns {boolean}
         */
        SmtpConfigurationsPager.prototype.hasNext = function () {
            return this._hasNext;
        };
        /**
         * Returns the next page of results by invoking listSmtpConfigurations().
         * @returns {Promise<EventNotificationsV1.SMTPConfiguration[]>}
         */
        SmtpConfigurationsPager.prototype.getNext = function () {
            return __awaiter(this, void 0, void 0, function () {
                var response, result, next;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!this.hasNext()) {
                                throw new Error('No more results available');
                            }
                            if (this.pageContext.next) {
                                this.params.offset = this.pageContext.next;
                            }
                            return [4 /*yield*/, this.client.listSmtpConfigurations(this.params)];
                        case 1:
                            response = _a.sent();
                            result = response.result;
                            next = null;
                            if (result && result.next) {
                                if (result.next.href) {
                                    next = (0, ibm_cloud_sdk_core_2.getQueryParam)(result.next.href, 'offset');
                                }
                            }
                            this.pageContext.next = next;
                            if (!this.pageContext.next) {
                                this._hasNext = false;
                            }
                            return [2 /*return*/, result.smtp_configurations];
                    }
                });
            });
        };
        /**
         * Returns all results by invoking listSmtpConfigurations() repeatedly until all pages of results have been retrieved.
         * @returns {Promise<EventNotificationsV1.SMTPConfiguration[]>}
         */
        SmtpConfigurationsPager.prototype.getAll = function () {
            return __awaiter(this, void 0, void 0, function () {
                var results, nextPage;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            results = [];
                            _a.label = 1;
                        case 1:
                            if (!this.hasNext()) return [3 /*break*/, 3];
                            return [4 /*yield*/, this.getNext()];
                        case 2:
                            nextPage = _a.sent();
                            results.push.apply(results, nextPage);
                            return [3 /*break*/, 1];
                        case 3: return [2 /*return*/, results];
                    }
                });
            });
        };
        return SmtpConfigurationsPager;
    }());
    EventNotificationsV1.SmtpConfigurationsPager = SmtpConfigurationsPager;
    /**
     * SmtpUsersPager can be used to simplify the use of listSmtpUsers().
     */
    var SmtpUsersPager = /** @class */ (function () {
        /**
         * Construct a SmtpUsersPager object.
         *
         * @param {EventNotificationsV1}  client - The service client instance used to invoke listSmtpUsers()
         * @param {Object} params - The parameters to be passed to listSmtpUsers()
         * @constructor
         * @returns {SmtpUsersPager}
         */
        function SmtpUsersPager(client, params) {
            if (params && params.offset) {
                throw new Error("the params.offset field should not be set");
            }
            this._hasNext = true;
            this.pageContext = { next: undefined };
            this.client = client;
            this.params = JSON.parse(JSON.stringify(params || {}));
        }
        /**
         * Returns true if there are potentially more results to be retrieved by invoking getNext().
         * @returns {boolean}
         */
        SmtpUsersPager.prototype.hasNext = function () {
            return this._hasNext;
        };
        /**
         * Returns the next page of results by invoking listSmtpUsers().
         * @returns {Promise<EventNotificationsV1.SMTPUser[]>}
         */
        SmtpUsersPager.prototype.getNext = function () {
            return __awaiter(this, void 0, void 0, function () {
                var response, result, next;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!this.hasNext()) {
                                throw new Error('No more results available');
                            }
                            if (this.pageContext.next) {
                                this.params.offset = this.pageContext.next;
                            }
                            return [4 /*yield*/, this.client.listSmtpUsers(this.params)];
                        case 1:
                            response = _a.sent();
                            result = response.result;
                            next = null;
                            if (result && result.next) {
                                if (result.next.href) {
                                    next = (0, ibm_cloud_sdk_core_2.getQueryParam)(result.next.href, 'offset');
                                }
                            }
                            this.pageContext.next = next;
                            if (!this.pageContext.next) {
                                this._hasNext = false;
                            }
                            return [2 /*return*/, result.users];
                    }
                });
            });
        };
        /**
         * Returns all results by invoking listSmtpUsers() repeatedly until all pages of results have been retrieved.
         * @returns {Promise<EventNotificationsV1.SMTPUser[]>}
         */
        SmtpUsersPager.prototype.getAll = function () {
            return __awaiter(this, void 0, void 0, function () {
                var results, nextPage;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            results = [];
                            _a.label = 1;
                        case 1:
                            if (!this.hasNext()) return [3 /*break*/, 3];
                            return [4 /*yield*/, this.getNext()];
                        case 2:
                            nextPage = _a.sent();
                            results.push.apply(results, nextPage);
                            return [3 /*break*/, 1];
                        case 3: return [2 /*return*/, results];
                    }
                });
            });
        };
        return SmtpUsersPager;
    }());
    EventNotificationsV1.SmtpUsersPager = SmtpUsersPager;
})(EventNotificationsV1 || (EventNotificationsV1 = {}));
module.exports = EventNotificationsV1;
